core/src/main/java/org/infinispan/topology/CacheTopology.java

```diff
@@ -223,4 +249,42 @@ public Integer getId() {
          return Collections.<Class<? extends CacheTopology>>singleton(CacheTopology.class);
       }
    }
+
+   /**
+    * Phase of topology change during rebalance. Using four phases guarantees these properties:
+    *
+    * 1. T(x+1).writeCH contains all nodes from Tx.readCH (this is the requirement for ISPN-5021)
+    * 2. Tx.readCH and T(x+1).readCH has non-empty subset of nodes (that will allow no blocking for read commands
+    *    && reading only entries node owns according to readCH)
+    *
+    * Old entries should be wiped out only after coming to the {@link #STABLE} phase.
```

Comment:
No, I don't think the added complexity is worth it, as the only real benefit would be in the tests that check the data container directly.